revamp makefiles

DieracDelta opened the pull request:
This PR addresses issue #108 . Notable changes include:
- Correcting nondeterminstic first run. This is done by replacing all occurrences of `compile && run` with `run` in all `run_wp*` scripts. The compilation is instead done in parallel before tests are run.
- `memory_samples` makefiles now include all targets.
- `make clean` can be run multiple times in a row. This is done by adding the `-f` flag everywhere.
- `cd wp && make` still reinstalls everything, even if it is already installed. This is nearly instantaneous. It's not clear to me how to tell, without directly diffing libraries, whether or 
- The PR also reorganizes a bunch of targets. Specifically:
   - clean now also cleans the `resources/sample_binaries/` directory
   - `wp`'s default target is `make install`
   - the ordering of `wp`'s `install` target goes as follows:
     - attempt to uninstall plugin
     - build library (makes use of dune and make cache)
     - reinstall library
     - build plugin (makes use of bapbuild and make cache)
     - reinstall plugin
   - the plugin and library are only rebuilt if there are changes to `src/*.ml` and `src/*.mli` files, and `wp.ml`, respectively.

fortunac commented on wp/plugin/Makefile:
```diff
@@ -1,22 +1,31 @@
+
 test: install save_project dummy
-	ocamlbuild -r -use-ocamlfind -pkgs 'bap,z3,bap_wp,oUnit,re' -tag 'warn(A-48-44),debug,thread' -I tests test.native
+	ocamlbuild -r -use-ocamlfind \
```
What happens if I call `make test` from the `plugin` directory but none of the sample binaries are compiled?

DieracDelta replied:
Good catch! I can move the `make` invocations involving sample binaries from the wp root makefile to this makefile, which should fix this.

fortunac commented on wp/resources/sample_binaries/cbat-multicompiler-samples/csmith/run_wp.sh:
```diff
@@ -19,4 +19,4 @@ run () {
 }
 
-compile && run
+run
```
I'm wondering if we should remove the
```
compile () {
  make
}
```
on these scripts since we won't be using it anymore. What are your thoughts?

DieracDelta replied:
Yes, that makes sense! +1 I'll remove them

fortunac commented on wp/plugin/Makefile:
```diff
@@ -1,22 +1,31 @@
+		-tag 'warn(A-48-44),debug,thread' -I tests test.native
 	./test.native
 
 verifier:
```
Should we update the target for the verifier since we know where we are copying the files to?

fortunac replied:
I'm also wondering if we should include removing the verifier API when we uninstall. Right now, even if we uninstall, that file is still located in the api path.

jtpaasch reviewed (changes requested):
* My main comment: I think the overarching goal that Chris had in mind is that, in the wp/plugin Makefile, we should have the install target with dependencies on build targets, and build targets with dependencies on source files. Something like this:
  * `install: build`
  * `build: $(PASS_NAME).plugin`
  * `$(PASS_NAME).plugin: $(SRC_FILES)`
Something like that (and something analogous for the bap_wp Makefile). With such a hierarchy of dependencies, things will only be built if the source files or targets change. The `install` target will always run, but it won't trigger any rebuilds if nothing is changed. I assume it's okay for `install` to run every time? In any case, I was under the impression that this is the way Chris wants these wp/plugin and bap_wp Makefiles written. Is that right, @ccasin? If so, I don't think it would be too hard to update what you have here to that. On the other hand, maybe you have good reasons why these Makefiles don't do it this way!

* The second point: I am under the impression that a number of actions in these Makefiles can be deleted. For example, I am under the impression that `dune install` will install the stuff in opam, in which case the `opam pin/install .` actions can be deleted? Likewise, I am under the impression that `dune clean` and `bapbuild -clean remove` will delete the `_build` directories and other build artifacts, in which case the various `opam pin remove/uninstall` and `rm -f` actions can be removed?  I've added a bunch of questions/comments along these lines, but if you have good reasons for why these extra build actions are there, maybe add a comment explaining why for future posterity?

* My third comment. This is something that maybe you do or maybe you don't want to change. I leave it up to @DieracDelta and @fortunac. I see `set -x` inside your shell scripts. I generally do not put such options inside shell scripts, because the person or automation system that runs your scripts may need different things. By leaving `set` directives out of your scripts, the person who calls your scripts then has the freedom to turn those options on/off as they see fit, e.g., by calling `/bin/bash -x /path/to/your/script.sh`, if they care about seeing that much debug output, or `/bin/bash /path/to/your/script.sh` if they don't. Maybe you would prefer to put `-x` in your test runner (presumably an action in a Makefile), and then delete all the `set -x` directives from your scripts.

ccasin commented on wp/Makefile:
```diff
@@ -1,25 +1,63 @@
+	@printf $(FMT_STR) "CLEANING PLUGIN"
+	$(MAKE) -C $(PLUGIN) clean
 
 reinstall: clean install
```
It's odd to me that the behavior of `reinstall` is different than doing `uninstall` and then `install`.  In particular `reinstall` will clean, but the `uninstall` target does not.  Maybe this should be:

```
reinstall: uninstall.plugin uninstall.lib install
```

And you should have to call `clean` if your really want to rebuild, to match the behavior of the other `install`/`uninstall` targets?

DieracDelta replied:
To make things more consistent, I've rearranged such that the behavior of reinstall matches between makefiles, and is equivalent to `uninstall install`.

ccasin commented on wp/Makefile:
```diff
@@ -1,25 +1,63 @@
+FMT_STR := "\u001b[31m%s\u001b[0m\n"
 
-.PHONY: install doc test clean reinstall
+.PHONY: test.lib test.plugin test all clean
```
I think basically all the targets in this file should be marked phony?

DieracDelta replied:
Agreed and changed!

ccasin commented on wp/lib/bap_wp/Makefile:
```diff
@@ -1,18 +1,32 @@
+SHELL=/bin/bash
 
-all: build
+.PHONY: test all install
```
also `clean`, `build`, perhaps others

ccasin commented:
@jtpaasch @DieracDelta Regarding JT's top level comment:  Please don't get hung up on "the way I want things".  The goal is to have Makefiles that (1) rebuild things when needed, (2) don't rebuild things that aren't out of date, and (3) clean and uninstall everything relevant when asked, so we're not constantly wiping our docker containers / switches because they are weirdly broken.

The standard way to do that is to have target names that are the filename of the thing being built and to have prerequisites on those targets that include the sources files that are inputs.  (i.e., what JT said).  This is what I'm familiar with, so it's what I suggested.  However, if you've accomplished the key goals in a different way, it's fine.

ccasin reviewed (commented):
I'm stopping this review because it seems like JT already has substantial feedback, but thought I'd still post a few nits I spotted.  Don't get hung up on any of this stuff, it's minor, and you may disagree.

DieracDelta commented on wp/Makefile:
```diff
@@ -1,25 +1,63 @@
-BAP_WP = lib/bap_wp
-WP = plugin
-SAMPLE_BINS = resources/sample_binaries
+SHELL=/bin/bash
```
I agree. This is inconsistent and there's not a reason behind it as far as I can tell. I'm going to change to simple expansion everywhere such that we are consistent. It seems like recursive expansion *can* be sneaky in the same way that pointers are. I'm not sure what the standard practice is.

DieracDelta commented on wp/lib/bap_wp/Makefile:
```diff
@@ -26,7 +40,22 @@ doc:
+	dune install
+# only needs to be added if does not already exist
+ifeq ($(IS_INSTALLED),false)
+	opam pin add -y .
```
You are right! We can definitely simplify these things per your top level comment. I'm removing all of the `opam *` and `rm -f` commands in favor of solely using dune.

DieracDelta commented on wp/lib/bap_wp/Makefile:
```diff
@@ -1,18 +1,32 @@
+
+all: install
+
+clean: uninstall clean.local
```
Hmm. I did a bit of investigation. We can remove the `rm -f` invocations. I'm also wondering how many of the `dune` invocations we need. Dune's manpage is not very helpful regarding their functionality. However if we break down the targets, `clean.local` invokes `dune clean`, and `uninstall` invokes `dune uninstall`. These commands *appear* to be doing different things:

- `dune clean` removes `_build` but leaves the library as installed when you look for it with `dune installed-libraries | grep bap_wp`.
- `dune uninstall` does not remove `_build`, but does remove `bap_wp` when you grep dune for it.

It looks like these are disjoint tasks, so we probably need them both? It's worth noting that the plugin builds even after `dune clean` runs, but not after `dune install` runs. I'm not 100% sure.

DieracDelta commented on wp/plugin/Makefile:
```diff
@@ -1,28 +1,55 @@
 
-.PHONY: all build install clean verifier
+BUILD := _build/build_unique_id
+SRC_FILES = $(wildcard */*.ml) $(wildcard */*.mli)
```
I did not realize that `**` searched recursively!

DieracDelta commented on wp/plugin/Makefile:
```diff
@@ -1,28 +1,55 @@
+ifneq ($(shell opam pin list | grep "bap_wp" | wc -l), 1)
+	$(error "bap_wp lib is not installed. Please install using relevant makefile.")
+endif
+	printf "SO YA MESSED WITH: " $(SRC_FILES)
```
This makes sense! It felt like a hack because I was included a temporary file used exclusively to figure out if the project needed to be rebuilt. You're right--we can use `wp.plugin` as the file to watch (and the name of the target). This makes things feel much better.

I added the error message to let the user know what went wrong. However, this is redundant; I'll remove it.

asdjfk I missed deleting that print statement. Was using it to debug changed files haha

DieracDelta commented on wp/plugin/Makefile:
```diff
@@ -1,28 +1,55 @@
 
-test: install
-	ocamlbuild -r -use-ocamlfind -pkgs 'bap,z3,bap_wp,oUnit,re' -tag 'warn(A-48-44),debug,thread' -Is "lib,tests" test.native
+build: $(BUILD)
```
Yes! This makes a lot of sense.

DieracDelta commented on wp/resources/sample_binaries/Makefile:
```diff
@@ -1,6 +1,7 @@
 CLEAN_DIRS = $(SAMPLE_DIRS:%=clean-%)
+MAKEFLAGS += -j$(shell nproc)
 
 .PHONY: all subdirs $(BUILD_DIRS)
```
I have no idea where that came from. Removing it doesn't seem to break anything.

DieracDelta commented on wp/resources/sample_binaries/Makefile:
```diff
@@ -1,6 +1,7 @@
 CLEAN_DIRS = $(SAMPLE_DIRS:%=clean-%)
+MAKEFLAGS += -j$(shell nproc)
 
 .PHONY: all subdirs $(BUILD_DIRS)
```
I don't think that the child targets are treated as PHONY. They're cached and not rebuilt every time (as far as I can tell). However, `make` *is* invoked on each of the subdirectories. I've removed the PHONY annotation and the directories still rebuild on change. I'm going to leave it like that since I can't think of a good reason to keep around PHONY.

ccasin replied:
I am also not an expert, but I thought the general rule was "any target whose name is the filename of the thing it builds should be marked PHONY".  It's entirely possible I'm wrong.

jtpaasch replied:
I seem to remember reading (I believe in the manual) that PHONY targets are not checked/tracked to see if they've been updated. In effect, something marked PHONY runs every time.

DieracDelta commented on wp/Makefile:
```diff
@@ -1,25 +1,63 @@
-	$(MAKE) -C $(WP) $@
+all: install
+
+install: uninstall.plugin install.lib build.plugin install.plugin
```
This makes sense. I've moved things around s.t. the dependencies work out and we can only have `install` rely on `uninstall.plugin` and `install.plugin`.

DieracDelta commented:
@jtpaasch @ccasin I've done my best to address the suggestions yall have made. Thanks again for the feedback!

Notably:

- I've rearranged the targets as JT suggested such that the `build` target relies on a file target that is what is the file being built.
- I've removed redundant invocations to `rm -f` and `opam`. This is a great catch JT! I think this marginally speeds things up.
- I've removed `set -x` from the test files. It makes sense to me to set that when invoking the script.
- I've copied JT's VIBES makefile style throughout these makefiles. I'm a fan and the style feels quite clean.

A couple of other things based on what chris mentioned:

- The default target will still reinstall everything when invoked. This should be fine since installation is fast, but building is slow.
- Make uses a cache (targets are rebuilt only when changes have been made).
- `clean` and `uninstall` remove all build artifacts and uninstall packages, respectively.

jtpaasch reviewed (approved):
Wow, this looks so understandable! The targets are simple and clean. Really nice. 

This looks great to me. I say merge 'er.

It's funny that you copied the VIBES Makefile, because it was me that copied you. When I made that file, I really was just trying to implement what you and Chris were saying as you worked on this ticket!

Now....in 6 months, I will have forgotten all the `make` stuff I learned here.

ccasin commented:
Agreed - merge away.
